### PR TITLE
BUG: fix IntegerArray astype with copy=True/False

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
xref https://github.com/pandas-dev/pandas/pull/34307#discussion_r443370993

Right now, we were not copying the mask correctly with `copy=True`, and also not when the data was actually copied (then the mask also needs to be copied, even with `copy=False`)

### CODE REVIEW

**Review by TomAugspurger — APPROVED** [review]
LGTM. The matplotlib test failure is unrelated and I think reported elsewhere.

Does this need a whatsnew, or is it only on master?

**Comment by jorisvandenbossche** [comment]
One additional API question that pops up: in numpy, `astype(.., copy=False)` seems to return itself if it doesn't need to copy (I am actually also relying on this in the `if data is self._data` check).  
But in the EA implementation, we always return a new array *object*, potentially with the same backing data. But we could in principle also do something like `if dtype == self.dtype and not copy: return self` (and that should already catch the one case where a copy can be avoided I think, and we could simply always copy in the other cases)

**Comment by jorisvandenbossche** [comment]
> Does this need a whatsnew, or is it only on master?

Not only master it seems, will add a whatsnew.

**Comment by TomAugspurger** [comment]
I don't have a strong opinion on returning a new object backed by the data vs. just returning `self`.

File: pandas/core/arrays/masked.py [hunk 1]
```diff
@@ -40,6 +40,17 @@ class BaseMaskedDtype(ExtensionDtype):
...
+    @classmethod
+    def construct_array_type(cls) -> Type["BaseMaskedArray"]:
+        """
+        Return the array type associated with this dtype.
+
+        Returns
+        -------
+        type
+        """
+        raise NotImplementedError
```
Comment (jreback) [inline pandas/core/arrays/masked.py:52]: should use AbstractMethodError

**Review by jreback — CHANGES_REQUESTED** [review]
lgtm very minor comment

File: pandas/core/arrays/masked.py (hunk 1 above)
Comment (jorisvandenbossche) [reply pandas/core/arrays/masked.py:52]: We actually use NotImplementedError in the base dtype class for this, so this is consistent with that (not fully sure why we do that though, there might be a reason)

**Comment by jorisvandenbossche** [comment]
Any preferences on what I do with `copy=False`? (returning a new object vs returning self)

**Comment by TomAugspurger** [comment]
Matching NumPy (returning the same object) seems OK.

**Comment by jorisvandenbossche** [comment]
Updated to return same object

File: pandas/core/arrays/masked.py (hunk 1 above)
Comment (jreback) [reply pandas/core/arrays/masked.py:52]: kk yeah we should actually change that as this is a better error message (but ok)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm, does this need a whatsnew? e.g. changed in 1.1?

**Comment by jorisvandenbossche** [comment]
Added a bug fix note

**Comment by jreback** [comment]
thanks @jorisvandenbossche

**Pull request merged**